lib/src/migration.dart

```diff
@@ -0,0 +1,81 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'realm_class.dart';
+import 'native/realm_core.dart';
+import './realm_object.dart';
+
+/// A [Migration] object is passed to you when you migrate your database from one version
+/// to another. It contains the properties for the Realm before and after the migration.
+/// After the migration is complete, [newRealm] will become the authoritative version of
+/// the file.
+class Migration {
+  final SchemaHandle _schema;
+
+  /// The Realm as it existed just before the migration. Since the models have changed,
+  /// this Realm can only be accessed via the dynamic API (i.e. [Realm.dynamic] and
+  /// [RealmObject.dynamic]).
+  final Realm oldRealm;
+
+  /// The Realm as it exists after the migration. Before the end of the callback, you need
+  /// to make sure that all relevant data has been migrated from [oldRealm] into [newRealm].
+  final Realm newRealm;
+
+  Migration._(this.oldRealm, this.newRealm, this._schema);
+
+  /// Finds an object obtained from [oldRealm] in [newRealm]. This is useful when you
+  /// are working with objects without primary keys and want to update some information
+  /// about the object as part of the migration.
+  T? findInNewRealm<T extends RealmObject>(RealmObject oldObject) {
```

Comment:
We can expose a more general-purpose API. My thinking here was that people would be iterating over the old objects and looking up their versions in the new realm as that's the conventional way of doing migrations. But that may be too prescriptive - I don't have strong opinions and would be fine either way. 

@nielsenko @desistefanova what are your thoughts here - do you prefer a more specialized API or a more general-purpose one?

Reply:
I prefer to know where I'm searching to find something in the new realm or in the old realm. findInNewRealm is OK for me. It could be also findInMigratedRealm.

Reply:
I think the behaviour of `findInNewRealm` is quite different from `newRealm.find` in that it does not require a primary key. If we re-worked `find` so that it could actually look up a realm object from another realm without using the primary key, then we can drop `findInNewRealm`. Otherwise I think we should keep it, and with it the name, since there are two realms on a `MIgration` and you can only use this, to look up in one of them.